core/src/avm2/globals/flash/geom/Vector3D.as

```diff
@@ -0,0 +1,146 @@
+package flash.geom {
+    public class Vector3D {
+
+        // TODO: does the order matter?
+        public var w:Number;
+        public var x:Number;
+        public var y:Number;
+        public var z:Number;
+
+        // TODO: and here?
+        public static const X_AXIS : Vector3D = new Vector3D(1, 0, 0);
+        public static const Y_AXIS : Vector3D = new Vector3D(0, 1, 0);
+        public static const Z_AXIS : Vector3D = new Vector3D(0, 0, 1);
+
+        public function Vector3D(x:Number = 0, y:Number = 0, z:Number = 0, w:Number = 0) {
+            this.x = x;
+            this.y = y;
+            this.z = z;
+            this.w = w;
+        }
+
+        public function get lengthSquared():Number {
+            return this.x * this.x + this.y * this.y + this.z * this.z;
+        }
+
+        public function get length():Number {
+            return Math.sqrt(this.lengthSquared);
+        }
+
+        public function toString():String {
+            return "Vector3D(" + this.x + ", " + this.y + ", " + this.z + ")";
+        }
+
+        public function add(a:Vector3D):Vector3D {
+            return new Vector3D(this.x + a.x, this.y + a.y, this.z + a.z);
+        }
+
+        public function subtract(a:Vector3D):Vector3D {
+            return new Vector3D(this.x - a.x, this.y - a.y, this.z - a.z);
+        }
+
+        public function incrementBy(a:Vector3D):void {
+            this.x += a.x;
+            this.y += a.y;
+            this.z += a.z;
+        }
+
+        public function decrementBy(a:Vector3D):void {
+            this.x -= a.x;
+            this.y -= a.y;
+            this.z -= a.z;
+        }
+
+        public function clone():Vector3D {
+            return new Vector3D(this.x, this.y, this.z, this.w);
+        }
+
+        public function copyFrom(sourceVector3D:Vector3D):void {
+            this.x = sourceVector3D.x;
+            this.y = sourceVector3D.y;
+            this.z = sourceVector3D.z;
+            // w is not touched!
+        }
+
+        public function equals(toCompare:Vector3D, allFour:Boolean = false):Boolean {
+            return this.x == toCompare.x && this.y == toCompare.y && this.z == toCompare.z && (!allFour || this.w == toCompare.w);
+        }
+
+        public function nearEquals(toCompare:Vector3D, tolerance:Number, allFour:Boolean = false):Boolean {
+            if (allFour)
+                // TODO seems like in this case, tolerance is just ignored???
```

Comment:
I don't think this is correct. Do you have any example for this?

Reply:
I was just as surprised as you are. But look at these test cases:

```
// (100, 200, 300, 400).nearEquals((100, 200, 350, 400), 10, true)
false
// (100, 200, 300, 400).nearEquals((100, 200, 350, 400), 100, true)
false

// (100, 200, 300, 400).nearEquals((100, 200, 300, 450), 10, true)
false
// (100, 200, 300, 400).nearEquals((100, 200, 300, 450), 100, true)
false
```

Sure, a positive case (returning `true`) could also be added...

Reply:
Okay, this is really heckin' weird. I smell a serious bug in FP here:
```

// (100, 200, 300).nearEquals((100, 200, 300), 0, true)
false
// (100, 200, 300).nearEquals((100, 200, 300), 10, true)
true
```

Reply:
I think I figured this out... :) See the relevant comment in my updated commit.